lib/ukalloc: Make palloc/pfree page-granular (+ document main public `ukalloc` API)

andreittr opened the pull request:
### Description of changes

Previously bbuddy's palloc/pfree implementations behaved like a page-order allocator, with the assumption of dealing only in power-of-2 numbers of pages instead of the page-granular API _implied_ by `ukalloc`. This mismatch between expected/assumed and actual behavior would lead to memory corruption and unpredictable crashes if bbuddy was called without following its undocumented and unasserted assumptions.

Fortunately, I have found no code in core Unikraft breaks these assumptions in a way to cause visible bugs ( :crossed_fingers: ), but this is (1) a ticking timebomb, and (2) severely limiting developent of code that uses page allocation more seriously.

The issue of the `ukalloc` API being undocumented was a leading cause of this situation, and so this PR starts off with documenting the core "public-facing" functions of `ukalloc`:
- object allocation (`malloc()` & friends, `free()`)
- page allocation (`palloc()` & `pfree()`)

Next, we need to consider that true page allocation semantics cannot be implemented on top of `malloc`&co (most notably due to page allocs being partially freeable), and thus the compat functions `uk_palloc_compat` and `uk_pfree_compat` are not correct and would break non-trivial page-allocating code if used.
Thus, these two functions are removed, and any malloc-only allocator will now _expliciitly_ signal that it cannot provide `palloc/pfree`, with the caller deciding whether `memalign` provides a satisfactory fallback or if an error is warranted.

The only users of palloc/pfree in the core is xen platform/driver code. I have checked these callsites along with @mogasergiu and they all would seem to work fine with memalign too (on xen at least). I've added explicit fallback logic to all of them.

Finally, this PR fixes bbuddy's palloc/pfree implementations to be truly page-granular, fully conforming to the now-documented ukalloc API.

### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [x] Updated relevant documentation.

### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]


### Additional configuration

N/A

andreittr committed:
lib/ukalloc: Document public alloc & free API

This change adds documentation to the core memory allocation and free
functions exposed by the ukalloc API, aiming to clarify ambiguous
behavior and make clear to callers the semantics under which these
operate. There is NO CHANGE in behavior, even when it is clearly wrong
(e.g., posix_memalign). Behavioral changes are left for future work.

Checkpatch-Ignore: REPEATED_WORD
Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
lib/ukallocbbuddy: Make palloc/pfree page-granular

Previously bbuddy's palloc/pfree implementations behaved like a
page-order allocator, with the assumption of dealing only in power-of-2
numbers of pages instead of the page-granular API required by ukalloc.

While the exact semantics of palloc/pfree were not made explicit until
a recent commit ("lib/ukalloc: Document public alloc & free API"),
function signatures made it extremely easy for callers to break bbuddy's
silent assumptions, causing memory corruption and unpredictable crashes.

This change fixes bbuddy's palloc/pfree implementations to be truly
page-granular, fully conforming to the ukalloc API.

Checkpatch-Ignore: MACRO_ARG_REUSE
Signed-off-by: Andrei Tatar <andrei@unikraft.io>

mogasergiu commented on lib/ukallocbbuddy/bbuddy.c:
```diff
@@ -291,65 +308,91 @@ static inline unsigned long num_pages_to_order(unsigned long num_pages)
+{
+	unsigned long v = (uintptr_t)ptr;
+
+	UK_ASSERT(IS_ALIGNED(v, __PAGE_SIZE));
```
Maybe `PAGE_ALIGNED` would have worked better here?
But then again, in the future it will probably be moved to a paging library. Even though you don't need to enable a library to use its headers, it would like an added dependency. Though I'd mention this. Maybe it's best left as is here with `IS_ALIGNED` indeed.

andreittr replied:
I think there's an argument to be made for Unikraft to provide (via a paging lib or otherwise) 2 kinds of paging API for different consumers (at least conceptually):
- "kernel" paging API: all the arch goodness, knowledge about page table levels, hugepages, PTE flags, the works; basically what `<uk/arch/paging.h>` does and is used by the little code that genuinely cares about paging at a deep level (e.g., `ukvmem`)
- "user" paging API: `PAGE_SIZE` just as a "unit of finest granularity at which we can map memory" -- this is used by _a lot_ of code that doesn't really care about the details of paging, rather just wants a "default chunk size" that matches neatly with the underlying hardware. `bbuddy` is a perfect example of this.

Like I said, this difference should only be in concept; in practice these can be in the same (or dependent) headers. But it's useful to make it clear _how_ a piece of code cares about pages.

michpappas replied:
There's no doubt it's useful, and as a matter of fact I've meant to add it for a while now, but chose to wait until we move paging into its own library as well as some other re-arch work that will help us determine how to make `__PAGE_SIZE` (which in `arm64` can be configured to values other than 4K for the smallest granule) available even when the paging library is not enabled.

mogasergiu commented on lib/ukallocbbuddy/bbuddy.c:
```diff
@@ -291,65 +308,91 @@ static inline unsigned long num_pages_to_order(unsigned long num_pages)
+	return (void *)alloc_ch;
+
+err_nomem:
+	uk_pr_warn("%p: Cannot handle palloc request of %lu: Out of memory\n",
```
Interesting, printed as a warning even though on error path :thinking:. Oh well this was like this even before the PR anyway...

andreittr replied:
We're in the context of just one allocator here (not even necessarily the system alloc); we can't know if an OOM condition here is trivial or catastrophic, and it's best to leave it to the caller to signal an error if warranted. IMHO even a warning is overkill, I'd've made it info at most, but I want to keep previous observable behavior similar.

mogasergiu replied:
Yes, true. I am not requesting a change. It's just funny that it seems to be something [recurring](https://github.com/unikraft/unikraft/issues/1505#issuecomment-2553487095) with bbuddy :sweat_smile:

mogasergiu commented on lib/ukallocbbuddy/bbuddy.c:
```diff
@@ -291,65 +308,91 @@ static inline unsigned long num_pages_to_order(unsigned long num_pages)
+	/* Trim off any extra pages */
+	slack = (1ULL << ord) - num_pages;
+	buf = CHUNK_END(alloc_ch, ord);
+	while (slack) {
```
Since you are even declaring subscope variables and it's quite lengthy, this looks like it could go into its own separate (inline) function. Something that would take `buf` and `slack` as arguments maybe?

andreittr replied:
Such a function would have a sig like `func(struct uk_bbpalloc *b, unsigned long *slackp, char **bufp)` since it needs `b` for context and has to iterate buf & slack in tandem (having caller iterate them separately is asking for bugs). Leaving aside my personal distaste for mixed in/out arguments, such a function would have a hard time arguing its separate existence.
Consider the very unusual and specific pre- and post-conditions under which this code operates (which would need to be explicitly documented & human-enforced before/after calls for a static inline):
- pre: `buf` points to the first byte _after_ the target memory (an otherwise "invalid" pointer, but very useful in this loop)
- pre: `b` must be in an otherwise invalid bbuddy state: target memory is marked free in bitmap but not present in freelists
- post: bbuddy state is valid if-and-only-if `func` returns with a slack of `0`; otherwise you _must keep calling it_ before you do anything else, otherwise everything breaks

IMHO these make it a case of when splitting off a function actually hurts understandability and resilience to human error.

mogasergiu commented on lib/ukallocbbuddy/bbuddy.c:
```diff
@@ -358,68 +401,76 @@ static void *bbuddy_palloc(struct uk_alloc *a, unsigned long num_pages)
-			    || to_merge_ch->level != order)
+	/* Might need to return the buffer in multiple chunks */
+	do {
+		/* release `np` pages starting at `buf` back into freelists */
```
Ditto about moving to a function the inner `while` + I am not a fan of subscope variable declaration unless we are talking macros (IMO, generally speaking, if you needed subscope declarations, there is a chance you could have split that part in another function) but is there any reason you are declaring them inside the `do { } while` instead of at the beginning? They are going to end up existing anyway. I won't insist on this though, I guess it could also be personal taste :upside_down_face: 

So then maybe we could split out that inner while loop into a static inline separate function that declares and manipulates freed_ch/ct. Something that receives as arguments `bp`, `buf` and `num_pages` and assumes `num_pages` is indeed a power of 2?
That could then be used by other functions as well maybe?

Maybe I am understanding the code wrong though? I see the inner loop handling power of two num_pages and the while loop splitting the maybe not power of two num_pages to accomodate for the inner loop. It looks like it could be two functions if I am not mistaken.

andreittr replied:
The main loop does roughly 3 things every iteration until it exhausts the buffer:
1. find a candidate first chunk to free -- `ptr_order(buf, npages_order(num_pages))` does all the heavy lifting, everything else is simple arithmetic
2. "merge with free buddies" -- this is what the inner loop does, but in practice it just kicks free buddies out of the freelist and updates `freed_ch`, `freed_ct`, and `ord` accordingly. No actual "merging" is done until step 3
3. "create new power-of-2 chunk & link into freelist + advance buffer" -- this is when the actual merge happens and pages are visibly freed

This tight interdependence between steps poses a few issues if one were to attempt splitting off into separate functions (see pre- & post-conditions in `palloc` as well):
- between the start of step 2 and the end of step 3, bbuddy is in an invalid state: pages are marked free but missing from freelists (much like the invalid state in `palloc`'s inner loop) => step 2 must _always_ be followed by step 3
- the loop in step 2 must always be executed until exhaustion, otherwise bbuddy will be in invalid state: "at rest" all pages must be maximally merged (otherwise the quick order-based size check in `palloc` is wrong)
- step 2 requires the initial value of `ord` to have been calculated _exactly_ as it is in step 1, otherwise the entire argument for why the target buddy, if free, has a valid head struct does not hold

Considering these, I make the same case as for `palloc` that splitting off would do more harm than good here.

mogasergiu commented on lib/ukallocbbuddy/bbuddy.c:
```diff
@@ -291,65 +308,91 @@ static inline unsigned long num_pages_to_order(unsigned long num_pages)
+		chunk_tail_t *spare_ct;
+
+		UK_ASSERT(spare_len <= slack * __PAGE_SIZE);
+		UK_ASSERT(1ULL << sord <= slack);
```
Maybe I am exaggerating, but I am getting anxious whenever I don't see paranthesis around bitwise ops in situations like these :sweat_smile:. Can we add them: `UK_ASSERT((1ULL << sord) <= slack);`

mogasergiu reviewed (commented):
Had a first look. Looks ok to me. Though I left some comments because I think you already know from past experience that I tend to not like subscope declarations as much :sweat_smile:. Alas, feel free to consider them as suggestions, I will not insist on them.

I will have another look tomorrow with a fresh mind and likely give my RB. Cheers :beers:!

michpappas commented on lib/ukallocbbuddy/bbuddy.c:
```diff
@@ -291,65 +308,91 @@ static inline unsigned long num_pages_to_order(unsigned long num_pages)
+		ord++;
+	/* We use >= as ord may have been set arbitrarily high by num_pages */
+	if (ord >= FREELIST_SIZE)
+		goto err_nomem;
```
Preserving the comment wouldn't hurt here.

```diff
+	/* Find the smallest order of free memory that satisfies the request */
	while (ord < FREELIST_SIZE && FREELIST_EMPTY(b->free_head[ord]))
```

michpappas commented on lib/ukallocbbuddy/bbuddy.c:
```diff
@@ -291,65 +308,91 @@ static inline unsigned long num_pages_to_order(unsigned long num_pages)
+	slack = (1ULL << ord) - num_pages;
+	buf = CHUNK_END(alloc_ch, ord);
+	while (slack) {
+		const size_t sord = npages_order(slack);
```
Nit: For consistency with the other vars you might as well rename `sord` to `spare_ord`. (also what is a `sord`  ⚔️)

michpappas commented on lib/ukallocbbuddy/bbuddy.c:
```diff
@@ -291,65 +308,91 @@ static inline unsigned long num_pages_to_order(unsigned long num_pages)
-	alloc_ch = b->free_head[i];
-	b->free_head[i] = alloc_ch->next;
+	/* Grab & unlink a chunk */
+	alloc_ch = b->free_head[ord];
```
For the sake of correctness and ease of review, both link and unlink would benefit from becoming inline functions or macros. We can also leave for a future iteration though, as it's not the only thing that could use a revamp.

andreittr replied:
Yeah, might even be worth switching over to [`<uk/list.h>`](https://github.com/unikraft/unikraft/blob/staging/include/uk/list.h) as it's clear we need doubly-linked lists here, and one fewer bespoke implementation is always a good thing. I'll consider it in the pass.

michpappas commented on lib/ukallocbbuddy/bbuddy.c:
```diff
@@ -358,68 +401,76 @@ static void *bbuddy_palloc(struct uk_alloc *a, unsigned long num_pages)
+
+		/* Try merging with existing free chunks */
+		while (ord < FREELIST_SIZE - 1) {
+			chunk_head_t *const targ = BBUDDY_BUDDY(freed_ch, ord);
```
Another nit: Maybe it's just me, but `targ` didn't make any sense (`t<something> arg`?) until I you said it stands for `target` (and now it seems obvious of course). As this implementation is complex enough already, let's use clear variable names. Consider whether `buddy` would be more a more suitable name, as you use that in the documentation above, or at least whether it's worth renaming to `target`.

andreittr replied:
Yeah, renaming to `buddy` makes sense, will do. The declaration will be very satisfyingly buddified :smile:  
```c
chunk_head_t *const buddy = BBUDDY_BUDDY(freed_ch, org);
```

andreittr commented:
Pushed changes addressing some comments & improving documentation.
Cosmetic & comment changes ONLY; no functionality diff.

michpappas commented on lib/ukallocbbuddy/bbuddy.c:
```diff
@@ -291,135 +308,207 @@ static inline unsigned long num_pages_to_order(unsigned long num_pages)
+			/* It is safe to deref this iff targ is marked free */
+			chunk_head_t *const targ = BBUDDY_BUDDY(freed_ch, ord);
+
+			/* Stop if buddy is not free or of wrong order */
```
I was about to say that a few words about why it's possible to get a buddy at a different (not wrong per se) order would help, but now I'm thinking it's paradoxical to request a buddy *of a given order* a few lines above with `BBUDDY_BUDDY(freed_ch, ord)` and in the very next line check if the return buddy is indeed of that order. I understand why it happens, but perhaps we should handle that case at the macro instead and return NULL?

andreittr replied:
The two lines (`BBUDDY_BUDDY` & `buddy->level` check) do conceptually different things, and the reasoning here goes as follows:
1. `BBUDDY_BUDDY` is pure arithmetic: it answers "I have a chunk of a particular order; what would the first page of my buddy chunk (i.e., the one to merge with to get a chunk of higher order) be?"
2. At this point we have a pointer to a "target" page, we know nothing about whether our "buddy chunk" actually exists or is free, just that if it were, it would start at this particular page. And that page could be either occupied or free
2a. If target page is occupied, there is no way our buddy could exist and be free, so we stop
3. If the target page is free, we know that is must be a chunk head (see reasoning in comment higher up); we still know nothing about _how big_ this target chunk is.
3a. If the target chunk is indeed of the same size as our chunk (i.e., same ord, read explicitly from the chunk header) then we know we found our buddy, it is free, and we can merge
3b. If the target chunk is of a different size, then it is _not_ our buddy yet -- it still needs to merge with someone else before merging with us. Thus we stop.

Considering the above flow, I don't think it would make sense to burden the `BBUDDY_BUDDY` macro with all 3 steps (which are required to correctly answer the question: "does my buddy at this order exist in whole and is free?"). _Maybe_ there's a case to be made for these 3 steps being put into a `static inline` :shrug: 

I think I should maybe make this logic flow clearer though, with a comment or smth, because indeed it is not obvious why we need to explicitly check the chunk level in addition to buddy arithmetic + checking if free.

michpappas replied:
Coming back to review this after a while I came up with exactly the same question (i.e. how is it possible for the buddy to be at the wrong order), so maybe that's a sign that there's something to improve here. 
I propose (1) renaming to  `BBUDDY_BUDDY_ADDR()` and (2) to elaborate on the process in the comment on line 474. I post a proposal, feel free to adapt:
```
               /* Try to merge with buddy at incremental orders.
                 * 
                 * At a given level, if the buddy is available to merge (see below)
                 * we create a merged chunk, and continue iterating to 
                 * the next level to see if we can merge into a larger chunk.
                 * Once we reach the max order, or we can't merge with buddy
                 * at the current order (see below), we break from the loop and 
                 * link the resulting chunk to the current order's freelist.
                 *
                 * Looking up the merge buddy involves 3 steps (order is important):
                 * 1. Compute ptr of freed_ch's buddy page at order `ord` (BBUDDY_BUDDY_ADDR)
                 *  - buddy may or may not be there (may have been broken down into smaller chunks).
                 * 3. Check whether said buddy page is free
                 *  - if allocated we can neither merge or trust its chunk_head_t to read level, so we stop
                 * 3. Read buddy chunk head to check size
                 *   - we know it's a valid chunk head because of comment above
                 */
```

mogasergiu reviewed (approved):
Reviewed-by: Sergiu Moga <sergiu@unikraft.io>

andreittr commented:
Rebased to current staging

andreittr committed:
drivers/xen: Fallback page allocation to memalign

Previously xen drivers would (correctly) use uk_palloc for page
allocations, implicitly relying on internal ukalloc fallbacks for
allocator that did not offer a palloc/pfree interface. These internal
fallbacks are no longer guaranteed, since page alloc semantics are
fundamentally different to malloc-like semantics.

Fortunately all xen drivers seem to deal fine with either palloc or
memalign semantics; this change makes this fallback behavior explicit.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
plat/xen: Fallback mm page allocation to memalign

Xen mm (correctly) uses uk_palloc for page allocations, implicitly
relying on internal ukalloc fallbacks for allocators that do not offer a
palloc/pfree interface. These internal fallbacks are no longer
guaranteed, since page alloc semantics are fundamentally different to
malloc-like semantics.

Fortunately mm can deal with either palloc or memalign; this change
makes this fallback behavior explicit.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
lib/ukalloc: Remove compatibility palloc/pfree

In recent clarifications of the ukalloc API it has been made clear that
the page allocation (palloc/pfree) has fundamentally different
semantics to object allocation (malloc&co/free), most notably the fact
that page allocations can be partially released.

Thus, correctly implementing palloc/pfree in terms of memalign/free is
no longer possible, and the existing compat functions would corrupt the
memory of any program exercising the palloc API in any serious capacity.

This change removes the palloc and pfree compatibility functions
entirely, meaning malloc-only allocators now explicitly do not support
page allocations. A possible fallback to e.g., memalign needs to be
considered and decided on the calling end.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr commented:
Pushed update addressing comments + the previously unconsidered issue that `uk_palloc/free_compat` can no longer be correctly implemented under the newly-clarified page alloc API. Details in the PR description.

andreittr commented:
Pushed update adding commit to fix type mismatch of `uk_ffsl` between x86 and ARM, leading to compile errors on the latter.

andreittr commented:
Pushed rebase to `RELEASE-0.19.0`.

andreittr commented:
Pushed rebase to current staging + changes & drop of 1 redundant commit after merging of #1652 .

michpappas commented on lib/ukallocbbuddy/bbuddy.c:
```diff
@@ -291,135 +308,218 @@ static inline unsigned long num_pages_to_order(unsigned long num_pages)
+	UK_ASSERT(FREELIST_ALIGNED(alloc_ch, ord));
+
+	/* Trim off any extra pages */
+	slack = (1ULL << ord) - num_pages;
```
For readability, consider renaming `slack` and `buf` to `spare_base` and `spare_page_count`, or similar, to be consistent with `spare_ord`, `spare_len`, and the rest of the variables used for the same context below.

andreittr replied:
Sure, with the caveat that `buf` is not the _base_ of the spare area, rather its _end_; will reflect this in name.

michpappas commented on lib/ukallocbbuddy/bbuddy.c:
```diff
@@ -291,135 +308,218 @@ static inline unsigned long num_pages_to_order(unsigned long num_pages)
+	slack = (1ULL << ord) - num_pages;
+	buf = CHUNK_END(alloc_ch, ord);
+	while (slack) {
+		const size_t spare_ord = npages_order(slack);
```
Please delcare at the function scope.

andreittr replied:
Why?! That would entirely prevent it from being `const`, uselessly expand its scope beyond where it makes any sense, and thus make code _less_ correct. [Coding conventions](https://unikraft.org/docs/contributing/coding-conventions#variable-definitions) say "Variables are defined at the beginning of the block. The block is separated from the code with a blank line." and this is indeed a block. No mention about "Local variables must be declared in the outermost scope possible", nor do I think that such a rule would ever make sense.

michpappas replied:
> [Coding conventions](https://unikraft.org/docs/contributing/coding-conventions#variable-definitions) say "Variables are defined at the beginning of the block. The block is separated from the code with a blank line." and this is indeed a block. No mention about "Local variables must be declared in the outermost scope possible", nor do I think that such a rule would ever make sense.

Not quite. It is in fact documented, although quite poorly (see the `num` variable in the example).

<img width="935" height="898" alt="image" src="https://github.com/user-attachments/assets/4b1019e7-b3a7-4c0e-9797-fa3dfe1a0787" />

Anyway, let's take this offline.

andreittr replied:
Aw crap, I completely missed the details in the example and have been having the wrong mental picture all along :facepalm: ; you're right, my bad.

michpappas commented on lib/ukallocbbuddy/bbuddy.c:
```diff
@@ -291,135 +308,218 @@ static inline unsigned long num_pages_to_order(unsigned long num_pages)
+	buf = CHUNK_END(alloc_ch, ord);
+	while (slack) {
+		const size_t spare_ord = npages_order(slack);
+		const size_t spare_len = BBUDDY_MASK(spare_ord);
```
Setting length to mask is quite unintuitive, perhaps it would be worth to alias this macro to a more meaningful name? Ideally we should rename `BBUDDY_MASK` too given that it's impossible to tell what it does.

michpappas commented on lib/ukallocbbuddy/bbuddy.c:
```diff
@@ -291,135 +308,218 @@ static inline unsigned long num_pages_to_order(unsigned long num_pages)
+	/* Might need to return the buffer in multiple chunks */
+	do {
+		/* release `np` pages starting at `buf` back into freelists */
+		size_t ord = ptr_order(buf, npages_order(num_pages));
```
Please declare at the function scope.

andreittr replied:
No, `ord` makes no sense outside a loop iteration, nor should its value mean anything across iterations; a declaration at function scope would make code less correct & readable not more.
I am very strict about minimizing variable scope, and if strongarmed, the only way I see to keep correctness is to make the loop body a `static inline` with ptr in/out args for `buf` and `num_pages`.

michpappas commented on lib/ukallocbbuddy/bbuddy.c:
```diff
@@ -291,135 +308,218 @@ static inline unsigned long num_pages_to_order(unsigned long num_pages)
+				freed_ch = buddy;
+			} else {
+				/* buddy is successor; merge upwards */
+				freed_ct = CHUNK_TAIL_PTR(buddy, ord);
```
While looking across the implementation I noticed that we populate the tail in different places but never actually make any use of it. So perhaps the whole `struct chunk_tail_st` / `struct chunk_tail_t` is redundant?

andreittr replied:
I think so; I hadn't removed it in the first pass, expecting it to be used _eventually_, but it really does seem that chunk tails are never referenced; will remove.

michpappas reviewed (changes requested):
Did another round, mostly cosmetic comments. Also before I forget, I believe the Commit Format Check in the CI fails because of the backticks in the PR's title. If you edit it won't be fixed, but just FYI for future PRs.

andreittr commented:
Pushed major update, addressing comments, and refactoring more of bbuddy.

andreittr commented:
Pushed cosmetic update to conform to checkpatch.

michpappas commented on lib/ukallocbbuddy/bbuddy.c:
```diff
@@ -49,24 +52,20 @@
-	chunk_head_t **pprev;
-	unsigned int level;
+struct chunk_head {
+	struct uk_hlist_node link;
```
Nice change to use `uk_hlist` here, but was there something wrong with `next`? Let's keep conventional naming, no need to unnecessarily obscure things.

andreittr replied:
`next` makes no sense as a field name here as a hlist link contains both a `.next` and a `.pprev` field (hlist is a DLL, and we do need that for quick removal out of freelists on buddy merging).
Previously yes, these names made sense as `chunk_head` contained the linked list pointers as raw fields within itself, wheras now with `hlist`, we need to use a single field of type `struct uk_hlist_node`, which itself has `.next` and `.pprev` fields.
Rather than naming things confusingly (but in keeping with tradition) I'd rather be clear. As example to contrast which of these look more sane:
```c
struct chunk_head *h;

h->next.next = next_elem;
h->next.pprev = &prev_elem;
```
or
```c
struct chunk_head *h;

h->link.next = next_elem;
h->link.pprev = &prev_elem;
```
(I know these ops are hidden behind macros/inlines, but that's no reason to introduce extra madness into one's code; plus, this is what people would see under a debugger).

Or, to answer the question strictly, yes, there was a lot wrong with `next`, as it was an explicit linked list pointer instead of a link struct field, as `uk_hlist` uses. Since we changed the meaning of the field, it makes sense to change its name to match.

michpappas replied:
Oh right, I see your point. Somehow my mind at that moment was hardwired to the previous state where `chunk_head_st` had `next` and `pprev` embedded.

michpappas commented on lib/ukallocbbuddy/bbuddy.c:
```diff
@@ -291,135 +296,187 @@ static inline unsigned long num_pages_to_order(unsigned long num_pages)
-	size_t i;
-	chunk_head_t *alloc_ch, *spare_ch;
-	chunk_tail_t *spare_ct;
+	const size_t spare_ord = npages_order(spare_pages);
```
This block is too noisy. Mixing code declarations with function calls etc is generally discouraged (https://unikraft.org/docs/contributing/coding-conventions#variable-definitions), we usually limit that to simple initializations (e.g. integer or `NULL`) if at all. Please consider a cleaner alternative similar to:

```diff
diff --git a/lib/ukallocbbuddy/bbuddy.c b/lib/ukallocbbuddy/bbuddy.c
index 53f478d7b..b942247a9 100644
--- a/lib/ukallocbbuddy/bbuddy.c
+++ b/lib/ukallocbbuddy/bbuddy.c
@@ -323,16 +323,20 @@ static inline size_t bbuddy_trimchunk(struct uk_bbpalloc *b,
                                      unsigned long spare_pages,
                                      char *spare_end)
 {
-       const size_t spare_ord = npages_order(spare_pages);
-       const size_t spare_len = BBUDDY_LEN(spare_ord);
-       char *spare_base = spare_end - spare_len;
-       struct chunk_head *spare_ch = (struct chunk_head *)spare_base;
+       struct chunk_head *spare_ch;
+       const size_t spare_ord;
+       const size_t spare_len;
 
+       spare_ord = npages_order(spare_pages);
+       UK_ASSERT(spare_ord < FREELIST_SIZE);
+
+       spare_len = BBUDDY_LEN(spare_ord);
        UK_ASSERT(spare_len <= spare_pages * __PAGE_SIZE);
+
        UK_ASSERT(BBUDDY_PAGES(spare_ord) <= spare_pages);
-       UK_ASSERT(spare_ord < FREELIST_SIZE);
 
        /* Populate chunk & link in */
+       spare_ch = (struct chunk_head *)(spare_end - spare_len);
        spare_ch->page_order = spare_ord;
        uk_hlist_add_head(&spare_ch->link, &b->free_head[spare_ord]);
 ```

andreittr replied:
~~Uhm, assigning to `const` variables after their definition is a no-no; the above code will not compile. I'm assuming that's an honest oversight.~~

~~That said, are you arguing that cosmetics are worth more than the extra safety against human screwups that immutable vars gives you? It's fine if you are (even if I disagree completely), but I want it explicit.~~

edit: Similar to subscope declarations, this was a case of me completely misreading the coding conventions and assuming degrees of freedom that weren't there. In light of this, the previous comments were out of line; standards aren't to be argued ad-hoc and I agree this code wasn't following them. Fixing in refactor/pass.

michpappas replied:
@andreittr many thanks for that. The part about `const` was of course an oversight of mine, so apologies. Review comments are always in good faith 🙃 Regarding the conventions: They are in many places quite unclear and could take a lot of improvement, so let's consider this an opportunity give them a revamp to avoid similar cases in the future.

michpappas commented on lib/ukallocbbuddy/bbuddy.c:
```diff
@@ -291,135 +296,187 @@ static inline unsigned long num_pages_to_order(unsigned long num_pages)
+static inline size_t bbuddy_freechunk(struct uk_bbpalloc *b, char *buf,
+				      unsigned long num_pages)
+{
+	size_t ord = ptr_order(buf, npages_order(num_pages));
```
Same here, please consider a layout that's easier to read, e.g.:
```diff
@@ -435,13 +439,19 @@ static inline int bbuddy_trymerge(struct uk_bbpalloc *b,
 static inline size_t bbuddy_freechunk(struct uk_bbpalloc *b, char *buf,
                                      unsigned long num_pages)
 {
-       size_t ord = ptr_order(buf, npages_order(num_pages));
-       const size_t np = BBUDDY_PAGES(ord);
-       struct chunk_head *freed_ch = (struct chunk_head *)buf;
+       struct chunk_head *freed_ch;
+       const size_t np;
+       size_t ord;
 
+       ord = ptr_order(buf, npages_order(num_pages));
        UK_ASSERT(ord < FREELIST_SIZE);
+
+       np = BBUDDY_PAGES(ord);
        UK_ASSERT(np <= num_pages);
 
+       freed_ch = (struct chunk_head *)buf;
+       UK_ASSERT(freed_ch);
+
        while (ord < FREELIST_SIZE - 1)
                if (bbuddy_trymerge(b, &freed_ch, ord))
                        ord++;
```

PS: Added an assertion for `freed_ch`, not sure if  a check against `buf` was intentionally omitted before.

andreittr replied:
> Added an assertion for freed_ch, not sure if a check against buf was intentionally omitted before.

Good catch, thanks; a NULL `obj`/`buf` should definitely be handled _somehow_. Whether gracefully or not I'm not sure; NULL is indeed an err value for `palloc`, so either way it should not be handled normally by `pfree`.
- On one hand, decreeing that `obj` must be non-NULL and crashing on non-compliance makes sense (and the [proposed wording](https://github.com/unikraft/unikraft/pull/1576/files#diff-c4f328020b85f182463f93103962fa30d9bd65d7e882c5784d75ee4f1322c94cR344) of pfree docs kinda says that already)
- OTOH, related APIs (e.g., `free()`) are explicitly NOPs when fed NULL, and people might expect symmetry / the principle of least surprise (although `palloc`/`pfree` is different enough to warrant being special :shrug: )

I'll go ahead and assert the argument, but let me know if you think NOPing out on NULL is the better call.

michpappas replied:
Hmmmm, I don't have a strong opinion. The important thing is to handle it somehow, so let's go for the suggested approach 👍🏼 We can always change it later if there is good reason.

michpappas commented on lib/ukallocbbuddy/bbuddy.c:
```diff
@@ -291,135 +296,187 @@ static inline unsigned long num_pages_to_order(unsigned long num_pages)
+static void bbuddy_pfree(struct uk_alloc *a, void *obj, unsigned long num_pages)
+{
+	struct uk_bbpalloc *const b = (struct uk_bbpalloc *)&a->priv;
+	char *buf = obj;
```
Any reason for the aliasing of `obj` to `buf`? Unless I'm missing something both can be used interchangeably in the function. Btw, I was saving this for a later revision with readability in mind, but since we're there, I always found `buf` an unfortunate name in the various places used in this file, perhaps `base` would be more intuitive? I also had a look at `malloc(3)` and friends, and they use `ptr`. If you think this is not for the present we can leave it for later, but at least let's remove the aliasing if it's indeed not needed.

andreittr replied:
The aliasing is only for doing pointer arithmetic on `buf` later on. The `pfree` signature requires us to take a `void *` (which we can't do arithmetic on directly), and manually casting
```c
obj = (char *)obj + freed * __PAGE_SIZE;
```
looks a bit noisy to me compared to a straight addition to `buf` :shrug: .

Re: naming, sure, `base` sounds good; I somewhat prefer it over `ptr` in this case as it more clearly expresses its role ("the start (base) of the buffer we're trying to divvy up into bbuddy chunks when freeing").

michpappas replied:
Okay fair enough, it's more of a matter of personal preference then.

michpappas commented on lib/ukallocbbuddy/bbuddy.c:
```diff
@@ -291,135 +296,187 @@ static inline unsigned long num_pages_to_order(unsigned long num_pages)
 
-		order++;
-	}
+	UK_ASSERT(a);
```
Would  `UK_ASSERT(b)` also be useful here?

andreittr replied:
:thinking: considering that `b` is always a constant away from `a` (a quick read of `ukalloc` says 14-16 pointers + the `uk_alloc_stats` (80 bytes) beyond `a`, depending on config). (note that `*b`'s contents are entirely embedded within `*a`, it's not a pointer field).

The first `UK_ASSERT(a)` guards us against the oopsie of someone passing a NULL allocator instance, an expected mistake and good to catch. A `UK_ASSERT(b)` would then only guard against passing `a` as `(void *)-208` (or -192/-128/-112 depending on config), which I feel is a bit too specific to be triggered by mistake, and thus I'm unsure whether it's worth guarding against at runtime.

michpappas replied:
> note that *b's contents are entirely embedded within *a, it's not a pointer field

Yup, that was it. I took it to be a pointer. It's all good now.

michpappas commented on lib/ukallocbbuddy/bbuddy.c:
```diff
@@ -291,135 +296,187 @@ static inline unsigned long num_pages_to_order(unsigned long num_pages)
-			if (allocated_in_map(b, (uintptr_t)to_merge_ch)
-			    || to_merge_ch->level != order)
-				break;
+	while (ord < FREELIST_SIZE - 1)
```
Maybe a brief comment for this while loop? To me it looks self explanatory, but then again I don't have a clear view because by now I am familiar with how it works. We also document the rest of the steps in this function, so I guess it wouldn't hurt adding a comment here too.

michpappas commented on lib/ukallocbbuddy/bbuddy.c:
```diff
@@ -291,135 +296,187 @@ static inline unsigned long num_pages_to_order(unsigned long num_pages)
+ * (b) is guaranteed to never happen on free, as buddy would need X to already
+ * be free in order to form a larger order
+ */
 
```
This newline is redundant.

michpappas commented on lib/ukallocbbuddy/bbuddy.c:
```diff
@@ -291,135 +296,187 @@ static inline unsigned long num_pages_to_order(unsigned long num_pages)
 
-	/* if the object is not page aligned it was clearly not from us */
-	UK_ASSERT((((uintptr_t)obj) & (__PAGE_SIZE - 1)) == 0);
+	/* Unlink buddy & continue to next order*/
```
"unlink and continue to next order" does not belong to this function, but to the caller. Which is of course not possible, as the caller doesn't know anything about the buddy. Now, I know this was made into static inline because of malicious compliance, but that's not the way forward. Maybe pass the candidate buddy's address to `bbuddy_trymerge()` instead, or merge this whole functionality into `bbuddy_freechunk()`?

andreittr replied:
Unlinking the buddy is actually completely done by `uk_hlist_del` below; the caller merely decides whether to move onto the next order or stop according to the return value of this function (thus `return 1` _technically_ tells the caller to continue to next order).
But I see how logic jumping from one func to another is a bit convoluted, I'll see if I can refactor these 3 functions better in another pass.

michpappas commented on lib/ukallocbbuddy/bbuddy.c:
```diff
@@ -291,135 +296,187 @@ static inline unsigned long num_pages_to_order(unsigned long num_pages)
-	UK_ASSERT((((uintptr_t)obj) & (__PAGE_SIZE - 1)) == 0);
+	/* Unlink buddy & continue to next order*/
+	uk_hlist_del(&buddy->link);
+	return 1;
```
The return value for success is a bit unconventional, convert into bool or reverse the logic?

andreittr replied:
Yeah, without context from the caller (which should never be required to understand a function) it's ambiguous what this means; will `bool`ify / refactor in pass.

michpappas commented on lib/ukallocbbuddy/bbuddy.c:
```diff
@@ -291,135 +296,187 @@ static inline unsigned long num_pages_to_order(unsigned long num_pages)
-			    || to_merge_ch->level != order)
-				break;
+	while (ord < FREELIST_SIZE - 1)
+		if (bbuddy_trymerge(b, &freed_ch, ord))
```
Perhaps should be called `buddy_canmerge()` instead? While reading this I expected that the function will also go ahead do the merging, if it can, which is not the case really.

andreittr replied:
`bbuddy_trymerge` does indeed do the merging; it's just that "merging" in this context means strictly only:
- [unlinking](https://github.com/unikraft/unikraft/pull/1576/files#diff-7a854c8df048714d0efcab86bedab15bf77f36afe24270e946fd1bdb578c14afR431) the corresponding buddy from its freelist
- (if buddy was a lower address than `freed_ch`) [updating `freed_ch`](https://github.com/unikraft/unikraft/pull/1576/files#diff-7a854c8df048714d0efcab86bedab15bf77f36afe24270e946fd1bdb578c14afR428) to point to the lower end

Any additional bookkeeping isn't really required until the last step when we insert the final, maximal chunk back into freelists.

But yeah, having merge logic split up like this makes it harder to grok; I'll try refactoring in the pass.

michpappas commented on lib/ukallocbbuddy/bbuddy.c:
```diff
@@ -291,135 +296,187 @@ static inline unsigned long num_pages_to_order(unsigned long num_pages)
 
-	size_t order = (size_t)num_pages_to_order(num_pages);
+	if (BBUDDY_ISHEAD(buddy, ord))
+		/* buddy is predecessor; merge downwards */
```
Perhaps one more sentence to clarify that if the buddy is not the head no action required as `free_ch` already points to head?

michpappas commented on lib/ukallocbbuddy/bbuddy.c:
```diff
@@ -49,24 +52,20 @@
-	unsigned int level;
+struct chunk_head {
+	struct uk_hlist_node link;
+	unsigned int page_order;
```
Also nice change to remove redundancy, it was one of the first things I noticed when I first read this 👍🏼

michpappas reviewed (changes requested):
@andreittr with the cleanup of the `chunk_head` and the use of `uk_hlist` it looks much better! 🚀  Went through it once again, didn't find any issues. Please bare with the last comments on readability, ready to RB after this round.

andreittr commented:
Pushed minor refactor & addressing comments.

michpappas reviewed (approved):
Good to go 🚀  @andreittr thanks for putting up with the elaborate review on this one.

Reviewed-by: Michalis Pappas <michalis@unikraft.io>
Approved-by: Michalis Pappas <michalis@unikraft.io>
